x-pack/plugins/alerting/server/alerts_client.ts

```diff
@@ -0,0 +1,141 @@
+  taskManager: TaskManager;
+  savedObjectsClient: SavedObjectsClient;
+  alertTypeRegistry: AlertTypeRegistry;
+}
+
+interface FindOptions {
+  options?: {
+    perPage?: number;
+    page?: number;
+    search?: string;
+    defaultSearchOperator?: 'AND' | 'OR';
+    searchFields?: string[];
+    sortField?: string;
+    hasReference?: {
+      type: string;
+      id: string;
+    };
+    fields?: string[];
+  };
+}
+
+interface CreateOptions {
+  data: Alert;
+  options?: {
+    migrationVersion?: Record<string, string>;
+  };
+}
+
+interface UpdateOptions {
+  id: string;
+  data: Alert;
+  options?: { version?: string };
+}
+
+export class AlertsClient {
+  private taskManager: TaskManager;
+  private savedObjectsClient: SavedObjectsClient;
+  private alertTypeRegistry: AlertTypeRegistry;
+
+  constructor({ alertTypeRegistry, savedObjectsClient, taskManager }: ConstructorOptions) {
+    this.taskManager = taskManager;
+    this.alertTypeRegistry = alertTypeRegistry;
+    this.savedObjectsClient = savedObjectsClient;
+  }
+
+  public async create({ data, options }: CreateOptions) {
+    const references = this.extractReferences(data);
+    const createdAlert = await this.savedObjectsClient.create<any>('alert', data, {
+      ...options,
+      references,
+    });
+    const scheduledTask = await this.scheduleAlert(createdAlert.id, data);
+    await this.savedObjectsClient.update('alert', createdAlert.id, {
+      scheduledTaskId: scheduledTask.id,
+    });
+    createdAlert.attributes.scheduledTaskId = scheduledTask.id;
+    return createdAlert;
+  }
+
+  public async get({ id }: { id: string }) {
+    return await this.savedObjectsClient.get('alert', id);
+  }
+
+  public async find({ options = {} }: FindOptions) {
+    return await this.savedObjectsClient.find({
+      ...options,
+      type: 'alert',
+    });
+  }
+
+  public async delete({ id }: { id: string }) {
+    const alertSavedObject = await this.savedObjectsClient.get('alert', id);
+    const removeResult = await this.savedObjectsClient.delete('alert', id);
+    await this.taskManager.remove(alertSavedObject.attributes.scheduledTaskId);
+    return removeResult;
+  }
+
+  public async update({ id, data, options = {} }: UpdateOptions) {
+    // Throws an error if alert type is invalid
+    this.alertTypeRegistry.get(data.alertTypeId);
+    // Re-create scheduled task if alertTypeId or interval changed
+    const currentSavedObject = await this.savedObjectsClient.get('alert', id);
+    const references = this.extractReferences(data);
+    const updatedObject = await this.savedObjectsClient.update<any>('alert', id, data, {
+      ...options,
+      references,
+    });
+    if (this.shouldRescheduleTask(currentSavedObject.attributes, data)) {
+      await this.taskManager.remove(currentSavedObject.attributes.scheduledTaskId);
+      const scheduledTask = await this.scheduleAlert(id, data);
+      await this.savedObjectsClient.update<any>('alert', id, {
```

Comment:
We would have a disconnect, either no task running or a new task created but saved object not updated. I'm not sure if there's a better option, changing the order may cause other problems (duplicate tasks, etc).

Reply:
Yeah, not having a true transaction here is an issue.  Not sure there's really a good solution to this problem.  Would it be possible to update the task with the new schedule (reuse the same ID)?  If we could do that, then the whole thing becomes a single operation.

Reply:
Just summarizing the offline conversation I had with you, we will drop support from changing `alertTypeId` which will remove the need of re-scheduling tasks. The interval change wasn't necessary to re-schedule a task since it's captured in the saved object.

Marking as resolved since code only updates saved object now.